Thrift #16

atkawa7 opened the pull request:
@josuegomes  @madduci @ras0219-msft Fails with build type debug but  works with release.

atkawa7 committed:
created Apache Thrift portfile

atkawa7 committed:
Create CONTROL

msftclas commented:
@atkawa7,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by Microsoft. We will now review your pull request.
_Thanks,
Microsoft Pull Request Bot_

atkawa7 committed:
fixed issue with binaries

atkawa7 commented:
@ras0219-msft  @madduci @josuegomes works with `vcpkg install --head thrift` but fails with `vcpkg install thrift`

KindDragon commented on ports/thrift/portfile.cmake:
```diff
@@ -0,0 +1,55 @@
+
+vcpkg_configure_cmake(
+    SOURCE_PATH ${SOURCE_PATH}
+    OPTIONS   -DWITH_SHARED_LIB=OFF -DWITH_STATIC_LIB=ON  -DBUILD_TESTING=off -DBUILD_JAVA=off -DBUILD_C_GLIB=off -DBUILD_PYTHON=off -DBUILD_CPP=on -DBUILD_HASKELL=off -DBUILD_TUTORIALS=off -DFLEX_EXECUTABLE=${FLEX_EXECUTABLE} -DBISON_EXECUTABLE=${BISON_EXECUTABLE}
```
`-DWITH_SHARED_LIB=OFF -DWITH_STATIC_LIB=ON`
I think you should dynamically set this options https://github.com/Microsoft/vcpkg/blob/14d9d0d60777146803892f7c9908c92b85867869/ports/c-ares/portfile.cmake#L11-L25

atkawa7 replied:
@KindDragon Thrift has no support for dynamic libraries on Windows

KindDragon replied:
Ok

ras0219-msft commented on ports/thrift/portfile.cmake:
```diff
@@ -0,0 +1,55 @@
+message(STATUS "download win flex bison")
+vcpkg_download_distfile(ARCHIVE
+    URLS "https://sourceforge.net/projects/winflexbison/files/win_flex_bison-latest.zip/download"
+    FILENAME "win_flex_bison-latest.zip"
```
This looks like the file will change in the future, which will cause the hash to no longer be valid and break the port.

We should use a stable version that will continue to work in the future.

ras0219-msft commented on ports/thrift/portfile.cmake:
```diff
@@ -0,0 +1,55 @@
+    HEAD_REF master
+)
+
+message(STATUS "download win flex bison")
```
I _think_ that flex and bison are *target architecture independent*, which means they would be good candidates to put into `vcpkg_find_acquire_program()`[1].

This makes it easy for other ports to get them, and it means they will be shared across different triplets and builds.


[1] https://github.com/Microsoft/vcpkg/blob/master/scripts/cmake/vcpkg_find_acquire_program.cmake

ras0219-msft commented on ports/thrift/portfile.cmake:
```diff
@@ -0,0 +1,55 @@
+    file(COPY ${EXES} DESTINATION ${CURRENT_PACKAGES_DIR}/tools)
+endif()
+
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
```
Duplicate of above line 43

atkawa7 committed:
Added Flex and Bison to vcpkg_find_acquire_program

atkawa7 committed:
Use vcpkg_find_acquire_program

-used vcpkg_find_acquire_program
-added warning  because  dynamic building not supported
-removed duplicate command

atkawa7 commented:
@ras0219-msft @KindDragon I have incorporated changes as requested

KindDragon commented on ports/thrift/portfile.cmake:
```diff
@@ -0,0 +1,40 @@
+include(vcpkg_common_functions)
+
+if (VCPKG_LIBRARY_LINKAGE STREQUAL dynamic)
+    message(STATUS "Warning: Dynamic building not supported. Building static.") # See note below
```
> See note below

Where is the note?

atkawa7 replied:
@KindDragon Added the note

atkawa7 committed:
Added note

bagong commented:
May I suggest to change the way flex and bison are acquired? The fact that the package contains both flex and bison is practical, but it makes the variable you chose unintuitive and PROGNAME useless. Would it maybe be better to actually split this up in two variables FLEX and BISON that both use the same download/hash/path? I've been experimenting with this as I need the same tools for a qtwebkit package. This seemed to be working well (it allows to use ${FLEX} and ${BISON} to access the executables, so the variables FLEX_EXECUTABLE and BISON_EXECUTABLE become unnecessary. That way it behaves like the other acquisitions. This code worked for me:

````
  elseif(VAR MATCHES "FLEX")
     set(PROGNAME win_flex)
     set(PATHS ${DOWNLOADS}/tools/winflexbison)
     set(URL "https://sourceforge.net/projects/winflexbison/files/win_flex_bison-2.5.9.zip/download")
     set(ARCHIVE "win_flex_bison-2.5.9.zip")
     set(HASH 9580f0e46893670a011645947c1becda69909a41a38bb4197fe33bd1ab7719da6b80e1be316f269e1a4759286870d49a9b07ef83afc4bac33232bd348e0bc814)
  elseif(VAR MATCHES "BISON")
     set(PROGNAME win_bison)
     set(PATHS ${DOWNLOADS}/tools/winflexbison)
     set(URL "https://sourceforge.net/projects/winflexbison/files/win_flex_bison-2.5.9.zip/download")
     set(ARCHIVE "win_flex_bison-2.5.9.zip")
     set(HASH 9580f0e46893670a011645947c1becda69909a41a38bb4197fe33bd1ab7719da6b80e1be316f269e1a4759286870d49a9b07ef83afc4bac33232bd348e0bc814)

````

One things worries me a bit though: why are the executables in these packages called win_bison and win_flex. Who would ever use them on a different OS??? What worries me is that I am sure some search routine or scripts will just look for `bison` and `flex` and fail... (I still like the idea of using the set of files bundled with the qt5-source, it seems the binaries are even a bit more recent (but it's not a archive):

    https://github.com/qt/qt5/tree/dev/gnuwin32

But if it is necessare to stick with the "official" download, and while you're at it, maybe also add gperf (which is already contained in the qt5-bundle)? ;)

````
  elseif(VAR MATCHES "GPERF")
     set(PROGNAME gperf)
     set(PATHS ${DOWNLOADS}/tools/gperf/bin)
     set(URL "https://sourceforge.net/projects/gnuwin32/files/gperf/3.0.1/gperf-3.0.1-bin.zip/download")
     set(ARCHIVE "gperf-3.0.1-bin.zip")
     set(HASH 3f2d3418304390ecd729b85f65240a9e4d204b218345f82ea466ca3d7467789f43d0d2129fcffc18eaad3513f49963e79775b10cc223979540fa2e502fe7d4d9)
````

Thanks for your consideration!

atkawa7 committed:
Remove unintuitive naming in vcpkg_find_acquire_program

atkawa7 committed:
Remove unnecessary Variables 

-Added vcpkg_find_acquire_program for flex and bison 
-Removed unnecessary variables

atkawa7 commented:
@bagong @ras0219-msft  I have added the new changes.

bagong commented:
Thanks @atkawa7. I've given this a bit more thought, sorry for coming up with more in a second request. But please consider it more a question. I observe that other packages are acquired with explicit versioning reflected in the path where the package is stored. E.g. ninja:

    elseif(VAR MATCHES "NINJA")
      set(PROGNAME ninja)
      set(SUBDIR "ninja-1.7.2")
      set(PATHS ${DOWNLOADS}/tools/ninja/${SUBDIR})

Admittedly you can easily query the version, but a little extra verbosity maybe makes the overall system more transparent? I am not passionate on this one, but find it a tad nicer than the bare rootname-folder (which I proposed myself earlier, sorry). Maybe there is a policy anyways?...

atkawa7 commented:
@bagong Thanks. I am not sure about this one. @ras0219-msft you thoughts?

bagong commented:
@atkawa7, yea, sorry, I just realize I actually misunderstood SUBDIR...

ras0219-msft commented:
The reason for `SUBDIR` is to handle upgrades.

When we need to upgrade the ninja version in the future (say there's some bugfix), we need to put it into a separate location so it doesn't conflict with the previously downloaded version.

Not all of the existing tools follow this pattern, but that's still OK. It just means when we upgrade the first time we'll need to move them into versioned subdirectories.

tldr; it would be a nice improvement but isn't critical 😸.

ras0219-msft commented:
I got some errors when building for x86-windows. Does this configuration work for anyone?

```
2017-05-26T21:32:36.6720692Z ===== D:\vcpkg-live-ci\_work\1\s\buildtrees\thrift\config-x86-windows-rel-err.log =====
2017-05-26T21:32:36.6720692Z CMake Error at C:/Program Files/CMake/share/cmake-3.8/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
2017-05-26T21:32:36.6876759Z   Could NOT find FLEX (missing: FLEX_EXECUTABLE)
2017-05-26T21:32:36.6876759Z Call Stack (most recent call first):
2017-05-26T21:32:36.6876759Z   C:/Program Files/CMake/share/cmake-3.8/Modules/FindPackageHandleStandardArgs.cmake:377 (_FPHSA_FAILURE_MESSAGE)
2017-05-26T21:32:36.6876759Z   C:/Program Files/CMake/share/cmake-3.8/Modules/FindFLEX.cmake:217 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
2017-05-26T21:32:36.6876759Z   compiler/cpp/CMakeLists.txt:22 (find_package)
```

I haven't investigated further yet -- it looks quite clear that `FLEX_EXECUTABLE` is being specified in `OPTIONS`...

atkawa7 commented:
@ras0219-msft try running with --head. There is an issue with the v 0.10.0

atkawa7 commented:
![image](https://cloud.githubusercontent.com/assets/5673828/26514178/17752db4-4225-11e7-8eb5-35b078a93533.png)

`error C2678: binary '==': no operator found which takes a left-hand ` 
This is the error I am getting when debug is turned on with the 0.10.0. However with head it works fine

ras0219-msft commented:
Doh! You already mentioned that :)

So, `vcpkg_from_github()` actually does support _only_ providing `HEAD_REF`. What do you think about just not having a stable ref for now, since it doesn't build? If you try to do `vcpkg install thrift` you should get an error that ["Package does not specify REF. It must built using --head."](https://github.com/Microsoft/vcpkg/blob/11d4a5c47edb75c8aaaa7d86bc2d351af8b32e28/scripts/cmake/vcpkg_from_github.cmake#L60).

Once a new stable release happens, we can easily add the REF/SHA back in. If we go this route, I'd recommend making the `Version:` be just "latest".

atkawa7 committed:
force users to use --head as latest stable fails to build

atkawa7 commented:
@ras0219-msft done :)

atkawa7 committed:
Removed hash

atkawa7 committed:
revert back to using sha

atkawa7 committed:
Removed sha and ref

atkawa7 committed:
Change version to latest

ras0219-msft committed:
[thrift] Small cleanup, added an unstable ref back in since the last stable doesn't build.

ras0219-msft commented:
Thanks for the PR and the quick change turnarounds 👍!

bagong commented:
There is a problem with this PR (in the acquire bison/flex part), sorry for noticing too late, I added the description to the issue:
https://github.com/Microsoft/vcpkg/issues/16#issuecomment-304525651
